### ISSUE TITLE
"The CSRF token is invalid." in tests after upgrading to 6.2.6 to fix CVE

### ISSUE THREAD

**Issue by J-roen** [symfony/symfony#49194]
### Symfony version(s) affected

6.2.6

### Description

After installing the fix for https://symfony.com/blog/cve-2022-24895-csrf-token-fixation, tests fail with the message "The CSRF token is invalid." I have only been able to reproduce this in tests, not (yet) when running the application in the browser.

@nicolas-grekas

### How to reproduce

1. Install the Symfony demo project: https://github.com/symfony/demo
2. Run `./bin/phpunit`. All tests should succeed.
3. Run `composer update`.
4. Run `./bin/phpunit`. Multiple tests should fail. Add `| grep CSRF` and you will find the error message.

### Possible Solution

I have no clue yet. My conclusion is that `$this->csrfTokenStorage->clear();` is the cause of this bug (see https://github.com/symfony/symfony/commit/5909d74ecee359ea4982fcf4331aaf2e489a1fd4#diff-0ff1412624a79146c346925f2407eb4783b144da38ddb369ca30e49d046fab70R59), but removing this is obviously not an option as it is the fix for the CVE.

### Additional Context

Docker one-liner:

```shell
docker run -it --rm composer bash -c " \
  composer create-project symfony/symfony-demo my_project; \
  cd my_project/; \
  ./bin/phpunit; \
  composer update; \
  ./bin/phpunit | head -n 6; \
  ./bin/phpunit | grep CSRF \
"

**Comment by Guite** [comment]
Not sure if it is the same issue but as it sounds similar I will add my observation here: 
The security fix removing CSRF tokens after a login seems to conflict with token based authentication: we've implemented an `AccessTokenHandler` and now `$form->isValid()` returns `false` because of a CSRF error. Possibly the same cause...

**Comment by nicolas-grekas** [comment]
@nicolasleborgne asks something similar in https://github.com/symfony/symfony/discussions/49188

> Am I wrong or since the security fix for CVE-2022-24894 we cannot use in combination http basic auth and csrf protection ? Tokens are cleared at each successful login and login occurred at each request so, when submitting a form, tokens do not match.

To me, all use cases that hit this were likely vulnerable to CVE-2022-24894, so we're going to have find another way to prevent CSRF, because the previous one didn't work. I'm writing "we" because I don't know if Symfony can provide something - doc or code - to cover this.

From afar, I suppose that when there is a session, the session should be used to NOT "login" at every request.

**Comment by MatTheCat** [comment]
Login from the session does not cause any issue because it is done with the `ContextListener`. However authenticators can ultimately cause a `LoginSuccessEvent` to be dispatched up to the `SessionStrategyListener` which will clear the CSRF token.

Symfony Demo’s tests authenticate using the `HttpBasicAuthenticator` on every request so when a form is submitted you’re authenticated again and the CSRF token is removed from the session.

**Comment by nicolas-grekas** [comment]
Good call: let's figure out how to fix the demo app, that will give us the food for thoughts we need here. Anyone up to give it a try?

**Comment by MatTheCat** [comment]
Using `KernelBrowser::loginUser` seems to do the trick :ok_hand: but it only fixes tests…

**Comment by gnutix** [comment]
I've just upgraded `symfony/security-(http|bundle)` to 6.0.20 and one of our Behat test fails because of "Invalid CSRF token". I've tried doing exactly what the test is doing by clicking in the application and it works, therefore the issue is only in the tests.

To authenticate the user, we use a custom "I am logged in with email XXX" step that sets an HTTP Header, caught by a custom Authenticator.

```
  Scenario: ...
    Given I am logged in with email "some@email.com"
    When I go to "/some/page"
    And I fill in "some_form_field" with "some_value"
    And I press "continue_button"
    # And print last response
    Then the url should match "/some/success/page"
```

The "url should match" check fails because we're back on /some/page, and printing the last response shows "Invalid CSRF token" error in the HTML content.

<details>
<summary>Here's the detailed code</summary>

```php
    /**
     * @Given I am logged in with email :identifier
     */
    public function iAmLoggedInWithEmail(string $identifier): void
    {
        $this->restContext->iAddHeaderEqualTo('Authorization', HeaderAuthenticator::AUTHORIZATION_HEADER_PREFIX . ' ' . $identifier);
    }
```

```php
#[When(env: 'dev'), When(env: 'test')]
final class HeaderAuthenticator extends AbstractAuthenticator
{
    public const AUTHORIZATION_HEADER_PREFIX = 'USER';

    public function __construct(
        private UserRepository $userRepository,
    ) {
    }

    public function supports(Request $request): ?bool
    {
        $header = (string) $request->headers->get('Authorization');

        return str_starts_with($header, self::AUTHORIZATION_HEADER_PREFIX);
    }

    public function authenticate(Request $request): Passport
    {
        $headerParts = explode(' ', (string) $request->headers->get('Authorization'));
        Assert::count($headerParts, 2, 'The authorization header should have two parts: the prefix and the user identifier.');

        $passport = new SelfValidatingPassport(new UserBadge(
            userIdentifier: $headerParts[1],
            userLoader: fn (string $identifier): UserInterface => $this->userRepository->loadUserByIdentifier($identifier),
        ));

        return $passport;
    }

    public function createToken(Passport $passport, string $firewallName): TokenInterface
    {
        return new TestToken(
            interactive: false,
            user: $passport->getUser(),
            firewallName: $firewallName,
            roles: $passport->getUser()->getRoles(),
        );
    }

    public function onAuthenticationFailure(Request $request, AuthenticationException $exception): Response
    {
        throw $exception;
    }

    public function onAuthenticationSuccess(Request $request, TokenInterface $token, string $firewallName): ?Response
    {
        // Let the request continue
        return null;
    }
}
```
</details>

**Comment by ghost** [comment]
I can confirm i have simmalair problem with version `v5.4.19`

**Comment by MatTheCat** [comment]
@nicolas-grekas would deferring the CSRF token removal on `kernel.response` still address the vulnerability?

**Comment by apemsel** [comment]
Confirmed on 5.4.20. When using x.509 authentication all forms stop working due to invalid CSRF token.

**Comment by apemsel** [comment]
Any advice what we should do until a proper solution is found if we have x509 and CSRF protection in production? Either pin symphony/security to a pre-patch version or disabled CSRF protection? The former might be less of a security risk?

**Comment by benr77** [comment]
`Upgrading symfony/security-http (v5.4.17 => v5.4.20)` gives me the exact same issue.

**Comment by MKCG** [comment]
I had the same issue, but actually it was caused by this commit https://github.com/symfony/http-kernel/commit/2a919968dfa24d995960d6ee96a69ccf3f79ba9b that resulted in the caching by the browser of some form pages containing a Set-Cookie header

**Comment by raziel057** [comment]
Same issue here following the https://symfony.com/doc/5.0/testing/http_authentication.html#using-a-faster-authentication-mechanism-only-for-tests documentation

**Comment by raziel057** [comment]
I fixed it by using the ``loginUser()`` strategy: https://symfony.com/doc/current/testing.html#logging-in-users-authentication

**Comment by MatTheCat** [comment]
Could someone confirm that https://github.com/symfony/symfony/pull/49319 fixes their issue?

**Comment by rivallinkil** [comment]
@MatTheCat #49319 fixes our issue (combination http basic auth and csrf protection) in unit test and dev env. 
Thanks !

**Comment by gnutix** [comment]
@MatTheCat @rivallinkil I'd like to try your patch out in our project, but I'm unsure how to apply the patch (as it seems to be impacting multiple Symfony components/bundles).

Would you mind providing some instructions on how to upgrade my project's dependencies or apply the patch locally ?

**Comment by florimondmanca** [comment]
@gnutix I asked the same question in #49319 and I was given these resources:

> You can follow https://symfonycasts.com/screencast/contributing/pr-reviewing (just skip the Creating a new Test App part as you already have an app to test on) to test a PR locally.

> see https://symfony.com/doc/current/contributing/code/pull_requests.html#use-your-branch-in-an-existing-project for the way to test your change locally.

**Comment by MatTheCat** [comment]
Sooo https://github.com/symfony/symfony/issues/49194#issuecomment-1413529705 is not a viable solution; sorry everyone..!

Maybe it is time to revive https://github.com/symfony/symfony/pull/33171 :thinking:

**Comment by nicolas-grekas** [comment]
Please try the fix in #49526 (it's for 5.4 but it shouldn't be that hard to adapt for 6.2)

**Comment by gnutix** [comment]
@nicolas-grekas Sorry to revive such an old issue, but my application runs on Symfony 6.4.23 now and I still get this issue. Not sure where's it at ?

Here's my Behat test :

```yaml
Feature: ...
  Scenario: Access "My parameters" page, proceed to wizard and be redirected to "My parameters" again.
    Given I am logged in with email "person1@domain.ch"
    When I go to "/resource/1/account-settings"
    And I go to "/security/2fa/method"
    And I fill in "verify_totp_code[authCode]" with "133737"
    And I press "continue_button"

    # @todo Fix this check, which is broken since Symfony 6.0.20, {@see https://github.com/symfony/symfony/issues/49194}
    # Then the url should match "/security/2fa/success"
    # When I follow "continue_button"
    # Then the url should match "/resource/1/account-settings"
```

And here's the interesting part of the response after the step `And I press "continue_button"` : 

```html
<form name="verify_totp_code" method="post" id="totpForm">
  <div class="d-flex justify-content-center justify-content-sm-start mt-3 is-invalid">
    <div>
      <input
        type="text"
        inputmode="numeric"
        name="verify_totp_code[authCode]"
        class="sms-code-input form-control is-invalid"
        placeholder="security.2fa.verification_code.code_input"
        autocomplete="off"
        maxlength="6"
      >
      <div class="form-error">
        <div class="invalid-feedback d-block">The CSRF token is invalid. Please try to resubmit the form.</div>
      </div>
    </div>
  </div>
  <input type="hidden" id="verify_totp_code__token" name="verify_totp_code[_token]" value="6cb50439f1fd7e822.sJ7p9hfFhLiPrbPlEdfhbWcnTORPOs-z5ZVg-gTmN3k.wfW5kFH1543In8qAXaOrKyZJeYAcfIrXj8cSrHOIbU3GyoSVc_DU4Nvc1A" />
</form>
```
